source/extensions/filters/network/connection_limit/connection_limit.cc

```diff
@@ -0,0 +1,96 @@
+#include "extensions/filters/network/connection_limit/connection_limit.h"
+
+#include "envoy/extensions/filters/network/connection_limit/v3/connection_limit.pb.h"
+
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ConnectionLimitFilter {
+
+Config::Config(
+    const envoy::extensions::filters::network::connection_limit::v3::ConnectionLimit& proto_config,
+    Stats::Scope& scope, Runtime::Loader& runtime)
+    : enabled_(proto_config.runtime_enabled(), runtime),
+      stats_(generateStats(proto_config.stat_prefix(), scope)),
+      max_connections_(PROTOBUF_GET_WRAPPED_REQUIRED(proto_config, max_connections)),
+      connections_(0),
+      delay_(PROTOBUF_GET_OPTIONAL_MS(proto_config, delay)) {}
+
+ConnectionLimitStats Config::generateStats(const std::string& prefix, Stats::Scope& scope) {
+  const std::string final_prefix = "connection_limit." + prefix;
+  return {ALL_CONNECTION_LIMIT_STATS(POOL_COUNTER_PREFIX(scope, final_prefix),
+                                     POOL_GAUGE_PREFIX(scope, final_prefix))};
+}
+
+bool Config::incrementConnectionWithinLimit() {
+  auto conns = connections_.load(std::memory_order_relaxed);
+  while (conns < max_connections_) {
+    if (connections_.compare_exchange_weak(conns, conns + 1, std::memory_order_release,
+                                           std::memory_order_relaxed)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void Config::incrementConnection() { connections_++; }
+
+bool Config::decrementConnection() {
+  auto conns = connections_.load(std::memory_order_relaxed);
+  while (conns > 0) {
+    if (connections_.compare_exchange_weak(conns, conns - 1, std::memory_order_release,
+                                           std::memory_order_relaxed)) {
+      return true;
+    }
+  }
+  return false;
```

Comment:
Added a warning log.
I don't think it's possible to return `false` theoretically, but it's good to put a log here in case some unexpected things happen.

Reply:
minor: add a prefix in the warning log to easily understand the context.

Reply:
Thanks @nitgoy, added.

Reply:
This seems like it should never happen unless there is a substantial logic bug. Should this just be an atomic decrement with an `ASSERT(connections_ > 0)` beforehand?

Reply:
Yes, it should never happen. I already added a warning log at the end of the function.
Added `ASSERT(connections_ > 0)` as well, but didn't change other code.
Because the `compare_exchange_weak` may return `false`, `while (conns > 0)` looks like a reasonable condition loop,
also, the function may return `false` in the test code.

Reply:
I don't think having this code makes sense since it can't happen. Just make it a simple atomic decrement, with an ASSERT before it, and make the function return void.

Reply:
Changed the `decrementConnection` function.